include/verilated_funcs.h

```diff
@@ -2339,6 +2339,60 @@ inline IData VL_VALUEPLUSARGS_INQ(int rbits, const std::string& ld, double& rdr)
 }
 extern IData VL_VALUEPLUSARGS_INN(int, const std::string& ld, std::string& rdr) VL_MT_SAFE;
 
+//======================================================================
+// bitHelper
+
+template <typename OutWordT, typename InWordT>
+void bitHelper(OutWordT* outWords, uint32_t outLsb, const InWordT* inWords, uint32_t inLsb,
```

Comment:
Add comments as to what this does please.  I'm still not convinced the VL_ASSIGN macros aren't better to use instead.

```suggestion
template <typename OutWordT, typename InWordT>
void vlBitHelper(OutWordT* outWords, uint32_t outLsb, const InWordT* inWords, uint32_t inLsb,
```

Reply:
Renamed and add comments to explain each argument.

I gave it a try to replace `vlBitHelper` by `VL_ASSIGNSEL_XX` series function. Following changes are necessary:

* Instead of using constant `vlBitHelper`, emit different `VL_ASSIGNSEL_XX` based on struct field width in `V3emitCHeader.cpp`.
* Extend `VL_ASSIGNSEL_XX` function to support additional `roffset` argument. `roffset` indicates skipping first `roffset` bits of input. Supporting `roffset` in `VL_ASSIGNSEL_WW` may take some efforts.
* Add and implement following additional function, which are needed to copy bits from `VlWide`.
```c++
static inline void VL_ASSIGNSEL_IW(int rbits, int obits, int lsb, CData& lhsr, WDataInP const rhs, roffset=0);
static inline void VL_ASSIGNSEL_IW(int rbits, int obits, int lsb, SData& lhsr, WDataInP const rhs, roffset=0);
static inline void VL_ASSIGNSEL_IW(int rbits, int obits, int lsb, IData& lhsr, WDataInP const rhs, roffset=0);
static inline void VL_ASSIGNSEL_QW(int rbits, int obits, int lsb, QData& lhsr, WDataInP const rhs, roffset=0);
```

I'm reluctant to the third change For it adds additional `VL_ASSIGNSEL_XX` functions with similar logic. Using a template is more concise and clear, although it may be a bit slower because of generality. But currently `vlBitHelper` is only used for user to send or see struct signals, I think it will not be some performance bottleneck.

Reply:
Ideally I'd like instead of printing code for you to create the code in the AST tree itself earlier, and Verilator would automatically create and optimize the C++ output code, but I was reluctant to suggest that work.  If you're open to it, that's one route and the best performing.

I suspect once these functions are available at least someone will use them every cycle, so they will likely become performance critical.  So better to assume so from the beginning, given we have better performing lower-level code already.

The conversion should be to create a deposit assign, from a selection, that's why you don't see the missing functions you mentioned being needed.  So e.g.

VL_ASSIGN_SEL_II(lhsr, VL_SEL_I[IWQ](rhs, from_rhs_bit)));
VL_ASSIGN_SEL_WW(lhsr, VL_SEL_W[IWQ](rhs, from_rhs_bit)));

The SEL can be suppressed when from_rhs_bit is zero.

Reply:
Updated comment:

Ideally I'd like, instead of printing code for you to create the code in the AST tree itself earlier, and Verilator would automatically create and optimize the C++ output code, but I was reluctant to suggest that work. If you're open to it, that's one route and the best performing.

I suspect once these functions are available, at least someone will use them every cycle, so they will likely become performance critical. So, better to assume so from the beginning, given we have better performing lower-level code already.

The conversion should be to create a deposit assign, from a selection, that's why you don't see the missing functions you mentioned being needed. So e.g.

VL_ASSIGN_SEL_II(lhsr, VL_SEL_I[IWQ](rhs, from_rhs_bit)));
VL_ASSIGN_SEL_WW(lhsr, VL_SEL_W[IWQ](rhs, from_rhs_bit)));

The inner SEL should be suppressed when from_rhs_bit is zero.

Likewise ASSIGN_SEL can be just ASSIGN when lhs_deposit_into_bit is zero.

Reply:
The new commit removes `vlBitHelper` function and adds a new batch of `VL_ASSIGNSEL_XX` functions
```c++
static inline void VL_ASSIGNSEL_II(int rbits, int obits, CData& lhsr, IData rhs,
                                   int roffset) VL_PURE;
static inline void VL_ASSIGNSEL_II(int rbits, int obits, SData& lhsr, IData rhs,
                                   int roffset) VL_PURE;
static inline void VL_ASSIGNSEL_II(int rbits, int obits, IData& lhsr, IData rhs,
                                   int roffset) VL_PURE;
static inline void VL_ASSIGNSEL_IQ(int rbits, int obits, CData& lhsr, QData rhs,
                                   int roffset) VL_PURE;
static inline void VL_ASSIGNSEL_IQ(int rbits, int obits, SData& lhsr, QData rhs,
                                   int roffset = 0) VL_PURE;
static inline void VL_ASSIGNSEL_IQ(int rbits, int obits, IData& lhsr, QData rhs,
                                   int roffset) VL_PURE;
static inline void VL_ASSIGNSEL_IW(int rbits, int obits, CData& lhsr, WDataInP const rhs,
                                   int roffset) VL_MT_SAFE;
static inline void VL_ASSIGNSEL_IW(int rbits, int obits, SData& lhsr, WDataInP const rhs,
                                   int roffset) VL_MT_SAFE;
static inline void VL_ASSIGNSEL_QW(int rbits, int obits, QData& lhsr, WDataInP const rhs,
                                   int roffset) VL_MT_SAFE;
static inline void VL_ASSIGNSEL_WW(int rbits, int obits, WDataOutP iowp, WDataInP const rwp,
                                   int roffset) VL_MT_SAFE;
```
Original `VL_ASSIGNSEL_XX` functions copy bits from lower bits of rhs to some offset of lhs, and lhs is always wider than rhs. This new batch of functions are opposite. They copy bits from some offset of rhs to lower bits of lhs , and rhs is always wider than lhs. It should has better peformance than general template.

Reply:
Thanks.  Please once manually grep your new tests' cpp files to make sure all the new functions are used, or, run coverage on verilator itself (nidust/code_coverage script) to check.

Reply:
Updated test. Run `nodist/code_coverage --test t_export_packed_struct.pl --test t_export_packed_struct2.pl` should see all `VL_ASSIGNSEL_XX` functions covered.